### DISCUSSION TITLE
Add wire:ref attribute to access component

### DISCUSSION THREAD

**Discussion by rw4lll** [livewire/livewire#5068] [3. Feature ideas]
After discuss with @joshhanley I've come to an idea, that probably it would be helpful to add new attribute **wire:ref** to the component. It should behave something like $ref attribute in Vue and provide and access to the component by its custom name. As a simple example, it would make possible to work with components from other js code on the page (e.g. legacy).
Something like:
```
<livewire:hello-world wire:ref="hello-world" />
...

var helloWorldComponent = Livewire.$ref('hello-world'); ->here we got livewire component instance
helloWorldComponent.set('property', 'value')
```
But I think also potentially there could be many other different scenarios. E.g. we could emit events targeted to specific $ref.

The topic was edited. Original question was:
> Hi everyone! When you use livewire components in a loop, its required to pass a wire key to avoid collisions. But can I somehow access the specific livewire component (on js side) by the wire key I provided with `:wire:key` attribute? As I see, its only possible by component id only on the moment? Probably it makes sense to provide such a method in js API?

**Comment by joshhanley** [comment]
@rw4lll yeah I think component ID is the only way at the moment. Can you share an example of what you are trying to do and what you would hope would work?

Thanks!

**Reply by rw4lll** [reply]
@joshhanley thanks for the response. Actually I've already refactored this part using different approach, but still think it could be a kinda useful feature. In my case I had a page with a list of products, where each product is represented by livewire component in foreach. Something like that:
```
@foreach($items as $itemId => $item)
  <livewire:item :item="$item" :wire:key="item-component-$itemId"/>
@endforeach
```
In my case after page load, I've requested an additional data for items from 3rd party providers, which is not stored on my side.
After I get the required data in json, my idea was to select necessary component using Livewire js API and mutate it using set() method:
```
var component = Livewire.components.findComponentByKey('item-component-$itemId');
component.set('additionalData', data);
```
The problem was that there is no such method to select component. You can do that only by using its unique uid which is randomly generated.

I could make a PR for that, but decided to ask beforehand.

**Reply by joshhanley** [reply]
@rw4lll thanks for checking!

Yeah I understand the usecase, but while keys are currently unique on a page, they shouldn't be and is a bug in morphdom, which causes all sorts of dom diffing issues.

We've been wanting to fix/ change it so keys are shallow, so they are only unique within a single element/ loop, the same as keys in Alpine, Vue, etc.

But this would be a breaking change, so that change is going to wait until V3.

Due to this though, that would mean what you are hoping to achieve/ PR won't work in V3.

Are you able to provide me some specifics and extra details on your exact use-case including concrete code of how and where it would be used, so I can get a better understanding of the requirements, so we can either find a way to get it to work, or see if we can come up with a PR that will work?

Thanks!

**Reply by rw4lll** [reply]
@joshhanley Ah, I understand. If in the future they will be changed, it makes no sense to extend it.

From my side I see this feature something like $ref attribute in Vue to be able to select specific component by its custom name and then perform some operations on it using js API. It can be especially useful when you starting with Livewire on existing project with legacy and need to provide some communication between old and new parts of code.

Probably it could be achieved by using new directive, something like :wire:ref?

**Reply by joshhanley** [reply]
@rw4lll ah yeah I see what you mean. Yeah some form of `wire:ref` could be interesting! Can you update the main title and content of this feature request to that? 🙂

**Reply by rw4lll** [reply]
@joshhanley done 🙂 . If such implementation could be interesting, I could prepare a PR for that.

**Reply by joshhanley** [reply]
@rw4lll I would wait to see what Caleb thinks of it first, before implementing it, to make sure he would be interested in something like that 🙂
